torch/csrc/autograd/engine.cpp

```diff
@@ -410,6 +441,22 @@ auto Engine::evaluate_function(FunctionTask& task) -> void {
       is_ready = true;
     }
 
+    // Switches to and sync with child's stream (if needed)
+    #ifdef USE_CUDA
+      // Sets child's device and stream
+      const int child_device = output.is_cuda() ? output.get_device() : - 1;
+      auto child_stream = next.function->input_metadata(next.input_nr).stream();
+      AutoGPUStream child_auto_gpu_stream{child_device, child_stream};
+
+      // Syncs GPU->GPU parent and child
+      if (parent_device != -1 
+      && child_device != -1 
+      && parent_stream != child_stream) {
+        auto_cuda_event.record(parent_stream);
+        auto_cuda_event.wait_on(child_stream);
+      }      
```

Comment:
This has a silent undocumented invariant which says that every op has a single device where all outputs live, and a single input device where all inputs live (and all inputs come from the same stream). While I think the device invariant is true (at least today - we should document it), the stream invariant doesn't seem to be guaranteed to hold.

Reply:
I don't think that's quire right, and this is a really interesting point of the current engine.

The way InputBuffer is written is highly suggestive that all input tensors to a function live on the same device. After all, you call .device() on the InputBuffer. This is somewhat misleading, however, and functions can actually receive outputs from and send outputs to multiple devices. Using DataParallel is an easy way to observe this behavior (and an easy way to trigger the metadata out of date issue above).

The current invariant in master is that each slot in the input buffer only contains tensors from the same device, and that cross-device synchronization is the responsibility of individual functions. I don't think this is ideal, and follow-ups to this PR could improve this invariant and take responsibility for these synchronizations, but that's outside our current scope.

Perhaps surprisingly, the addition of streams here don't change this invariant. Metadata is per slot, so each slot now has a device and a stream. Even if a function creates outputs on multiple streams on the same device, their metadata slot will still only have one stream. This might be considered a limitation of the current design. As long as these tensors are available before backward() is called, then, they will operate correctly. It is the caller's responsibility (just like today) to ensure synchronization across multiple streams in forward.

Does that make sense? Maybe hit me up on Slack if we need to get deeper into the details.